rust/src/tx_builder.rs

```diff
@@ -175,110 +181,244 @@ impl TransactionBuilder {
-                for output in outputs.iter().rev() {
-                    let mut added = Value::new(&Coin::zero());
-                    let mut needed = output.amount.clone();
-                    while added < needed {
-                        if available_inputs.is_empty() {
-                            return Err(JsError::from_str("UTxO Balance Insufficient"));
+            CoinSelectionStrategyCIP2::LargestFirstMultiAsset => {
+                // indices into {available_inputs} for inputs that contain {policy_id}:{asset_name}
+                let mut available_indices = (0..available_inputs.len()).collect::<Vec<usize>>();
+                // run largest-fist by each asset type
+                if let Some(ma) = output_total.multiasset.clone() {
+                    for (policy_id, assets) in ma.0.iter() {
+                        for (asset_name, asset_amount) in assets.0.iter() {
+                            self.cip2_largest_first_by(
+                                available_inputs,
+                                &mut available_indices,
+                                &mut input_total,
+                                &mut output_total,
+                                |value| value.multiasset.as_ref()?.get(policy_id)?.get(asset_name))?;
                         }
-                        let (new_input_total, new_output_total, new_added, new_needed, input) = add_random_input(self, &mut rng, &mut available_inputs, &input_total, &output_total, &added, &needed)?;
-                        input_total = new_input_total;
-                        output_total = new_output_total;
-                        added = new_added;
-                        needed = new_needed;
-                        associated_inputs.entry(output.clone()).or_default().push(input);
                     }
                 }
-                if !available_inputs.is_empty() {
-                    // Phase 2: Improvement
-                    for output in outputs.iter_mut() {
-                        let associated = associated_inputs.get_mut(output).unwrap();
-                        for input in associated.iter_mut() {
-                            let random_index = rng.gen_range(0..available_inputs.len());
-                            let new_input = available_inputs.get_mut(random_index).unwrap();
-                            let cur = from_bignum(&input.output.amount.coin);
-                            let new = from_bignum(&new_input.output.amount.coin);
-                            let min = from_bignum(&output.amount.coin);
-                            let ideal = 2 * min;
-                            let max = 3 * min;
-                            let move_closer = (ideal as i128 - new as i128).abs() < (ideal as i128 - cur as i128).abs();
-                            let not_exceed_max = new < max;
-                            if move_closer && not_exceed_max {
-                                std::mem::swap(input, new_input);
```

Comment:
I think this part was broken before anyway. Maybe that's why that phase 3 was necessary despite adding in the input fee to the required amount in phase 1 before? The inputs in `associated_inputs` would be distinct copies from the ones in the builder as they are added above as a copy.